stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
Great, I forgot about -O3. Now things got somewhat more interesting:

$ g++ main.cpp `pkg-config eigen3 --cflags --libs` -O3
$ ./a.out
start
4
11
4
0

$ g++ main.cpp -DEIGEN_NO_DEBUG `pkg-config eigen3 --cflags --libs` -O3
$ ./a.out
start
0
0
0
0

There might not be much meaning in creating this pull request after all...


On 03/03/2016 19:59, Bob Carpenter wrote:
>
> In stan/math/prim/mat/fun/value_of_rec.hpp 
> <https://github.com/stan-dev/math/pull/250#discussion_r54963577>:
>
> > -      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
> > -      for (int j = 0; j < M.cols(); j++)
> > -        for (int i = 0; i < M.rows(); i++)
> > -          Md(i, j) = value_of_rec(M(i, j));
> > -      return Md;
> > +      Eigen::Matrix<double, R, C>
> > +        result(M.rows(), M.cols());
> > +
> > +      double * data_r = result.data();
> > +      const T * data_m = M.data();
> > +
> > +      int S = M.size();
> > +      for (int i=0; i < S; i++)
> > +        data_r[i] = value_of_rec(data_m[i]);
> > +      return result;
> > +    }
>
> Are you using optimization |-O3| here? We only care about the 
> |EIGEN_NO_DEBUG| case for the Stan language (we do our own bounds 
> checking to give users informative messages), but the math library may 
> get used with the debug on.
>
> I'm guessing the fact that indexing rows and cols vs. size was the 
> same time means that the integer arithmetic used to compute the index 
> memory position is easily parallelized.
>
> But now I'm curious what extra work the indexing is doing or how it's 
> being defeated; this wouldn't be curious at -O1 where the inlining of 
> the call to the indexing function wouldn't happen.
>
> and now that I think about it, perhaps not even surprising. But 
> integer arithmetic (what you need to compute position of |a(i,j)|) is 
> easily parallelizable
> I think the indexes are reversed in (1) in your description.
>
> That's a lot of time spent bounds checking.
>
> Did the CLFAGS include
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/stan-dev/math/pull/250/files#r54963577>.
>